sklearn/cross_validation.py

```diff
@@ -332,6 +332,126 @@ def __len__(self):
         return self.n_folds
 
 
+def disjoint_label_folds(labels, n_folds=3):
+    """Creates folds where a same label is not in two different folds.
+    
+    Parameters
+    ----------
+    labels: numpy array, shape (n_samples,)
+        Contains an id for each sample.
+        The folds are built so that the same id doesn't appear in two different folds.
+    
+    n_folds: int, default=3
+        Number of folds to split the data into.
+        
+    Returns
+    -------
+    folds: numpy array of shape (n_samples, )
+        Array of integers between 0 and (n_folds - 1).
+        Folds[i] contains the folds to which sample i is assigned.
+        
+    Notes
+    -----
+    The folds are built by distributing the labels by frequency of appearance.
+    The number of labels has to be at least equal to the number of folds.
+    """
+    labels = np.array(labels)
+    unique_labels, labels = np.unique(labels, return_inverse=True)
+    n_labels = len(unique_labels)
+    if n_folds > n_labels:
+        raise ValueError(
+                ("Cannot have number of folds n_folds={0} greater"
+                 " than the number of labels: {1}.").format(n_folds, n_labels))
+    
+    # number of occurrence of each label (its "weight")
+    samples_per_label = np.bincount(labels)
+    # We want to distribute the most frequent labels first
+    ind = np.argsort(samples_per_label, kind='mergesort')[::-1]
+    samples_per_label = samples_per_label[ind]
+
+    # Total weight of each fold
+    samples_per_fold = np.zeros(n_folds)
+
+    # Mapping from label index to fold index
+    label_to_fold = np.zeros(len(unique_labels))
+    
+    # While there are weights, distribute them
+    # Specifically, add the biggest weight to the lightest fold
+    for label_index, w in enumerate(samples_per_label):
+        ind_min = np.argmin(samples_per_fold)
+        samples_per_fold[ind_min] += w
+        label_to_fold[ind[label_index]] = ind_min
+    
+    folds = label_to_fold[labels]
+
+    return folds
+
+
+class DisjointLabelKFold(_BaseKFold):
+    """Creates K approximately equilibrated folds.
```

Comment:
I would not put emphasis on balancing but rather on mutual label exclusion:

```python
class DisjointLabelKFold(_BaseKFold):
    """K-fold iterator variant with non-overlapping labels.

    The same label will not appear in two different folds  (the number of
    labels has to be at least equal to the number of folds).

    The folds are approximately balanced in the sense so that the number of
    distinct labels is approximately the same in each fold.
    """
```